Iicruz/rpc

ii-cruz opened the pull request:
- Introduces logs events to the blockchain. The Logs are aggregated by block in a struct called BlockLog.  Logs from the transactions are included as TransactionLogs, thus aggregated by transaction (TransactionLog struct). Inherent logs are included in the blocklog separately from the remaining logs.
- Adds the event stream subscription for BlockLogs on the rpc server. The logs can be subscribed to and filtered by addresses or log types.
- Adds block height and hash to the head_hash_subscribe, staker_by_address, and get_validator_by_address
- Adds the option to subscribe to full blocks (head_subscribe)

## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.

ii-cruz committed:
Block height and hash added to get functions
added head_subscribe fn for getting the full blocks

ii-cruz committed:
Notification streams for getting new block and validators on the election blocks

ii-cruz committed:
logs notifications added to the rpc and the blockchain.
adapted the blockchain transaction units to include log events.
rpc subscription based on addresses and types of log
events introduced.

ii-cruz committed:
serde changes

ii-cruz committed:
unit test fixing of the commit and rever inherent + formating fixing

sisou commented on rpc-server/src/dispatchers/blockchain.rs:
```diff
@@ -468,55 +525,57 @@ impl BlockchainInterface for BlockchainDispatcher {
+        Ok(stream
+            .map(move |hash| {
+                let blockchain_rg = blockchain.read();
+                get_block_by_hash(blockchain_rg.deref(), &hash, include_transactions)
+                    .map_err(|_| hash)
```
Does this, and the result type `Result<Block, Blake2bHash>` mean that this stream can _both_ yield a block object _and_ a hash string?

ii-cruz replied:
You are correct. So this function uses the head_hash_subscribe to have a stream of new block hashs and then fetches the block using that hash. In case there is a new block committed, the new hash is received over the stream. If the block was reverted in the meantime, the get_block_by_hash fails. In that case, we are sending the hash as a return value. 

The reason for this behavior is that serde is not able to serialize something like Result<Block, Error>. The alternative would be to filter out the event altogether. That could work too, I'm not seeing any issues with it.

sisou commented on rpc-server/src/dispatchers/blockchain.rs:
```diff
@@ -529,4 +588,152 @@ impl BlockchainInterface for BlockchainDispatcher {
+                            is_tx_logs_related_to_any_addresses(&tx_log, &addresses)
+                        });
+                        if !inherent_logs.is_empty() || !tx_logs.is_empty() {
+                            block_log = Some(BlockLog::AppliedBlockLog {
```
This is in the `BlockLog::RevertBlockLog` matching block, so I think:
```suggestion
                            block_log = Some(BlockLog::RevertBlockLog {
```

ii-cruz replied:
You are 100% correct. Thank you!

sisou commented on rpc-server/src/dispatchers/blockchain.rs:
```diff
@@ -529,4 +588,152 @@ impl BlockchainInterface for BlockchainDispatcher {
+                        inherent_logs.retain(|log| contains_log_type(&log, &log_types));
+                        tx_logs.retain(|tx_log| tx_logs_contains_any_log_type(&tx_log, &log_types));
+                        if !inherent_logs.is_empty() || !tx_logs.is_empty() {
+                            block_log = Some(BlockLog::AppliedBlockLog {
```
Same here, this is in the `RevertBlockLog` match block:
```suggestion
                            block_log = Some(BlockLog::RevertBlockLog {
```

ii-cruz committed:
Update rpc-server/src/dispatchers/blockchain.rs

Co-authored-by: Sören Schwert <hello@soerenschwert.de>

ii-cruz committed:
Update rpc-server/src/dispatchers/blockchain.rs

Co-authored-by: Sören Schwert <hello@soerenschwert.de>

sisou commented:
@ii-cruz I just noticed another problem: In the block subscription stream, all results pushed over the websocket are wrapped in a `Ok {}` result, e.g.

```
{
  Ok: {
    hash: '57fa19973988afa678beba2d48798fe26f54264888ae7a60f9a16b0ba3f23b39',
    size: 315,
    batch: 9,
    epoch: 3,
    version: 1,
    number: 283,
    view: 0,
    timestamp: 1656434106296,
    parentHash: '884a89736fb6518a8f3f8ba2ae2865ac0f3b852f98a56ea410f8dd1bb5f0f187',
    seed: { signature: [Array] },
    extraData: '',
    stateHash: '39ea52d1c50e6768916f54460c24df39d18bc7a260d20a98d19c5dcafef4e3d1',
    bodyHash: '11da6d1f761ddf9bdb4c9d6e5303ebd41f61858d0a5647a1a7bfe089bf921be9',
    historyHash: '81e47a19e6b29b0a65b9591762ce5143ed30d0261e5d24a3201752506b20f15c',
    type: 'micro',
    producer: {
      slotNumber: 143,
      validator: 'NQ69 MH90 3M5H DFNU 8E84 53FX JSY6 XYQS N2CE',
      publicKey: '000b17927b6293cd141488086073077392386c752c8d990340313ff7730e24497938a088c3694c827501e6474b3c288a075f919cc5eebde20a812cf2b475da77dc961eedcc0216bf53010e2df22dd19a20fe62f0416323a966a0b78ceb465d009700ed01a91f339a087580c86bf6b68e1c71d7bbebd65c5b5fca0b36c6f9bfcbb03cb4746d2164738dcbc9f825ac2edb6178a2b43665e615ad38fe9516ae671cc5c31eab5102c92a111ca309d3a9fc88f64a0b56959db881d663bc085fc400668e719eb7a8ef674add0520f16f28b31864953c7899fb458d491486aea8e6d30ddb107e522e9409531e7d5aa877e6ef9a80ce88c176f6abc690daed3e07af5a88700bebb6c3833a198f13121f2ebf426560cbe933857f0ee522c8e430ec'
    },
    forkProofs: [],
    justification: {
      signature: '5d875768ffd066ec41522cfe15cb5f8ba4e5c1d049e6256f93d3d8a7323a1f52fb010b4fe1f109fe48dd46d24044b64a7a7e2248f10316f4950bb32e20cbe90f'
    }
  }
}
```

Please unwrap the data before sending it over the wire.

sisou committed:
Unwrap enums when sent as JSON

sisou commented:
@paberr The logs are sent as JSON with snake_cased names. Can we get the fields of enum structs to also be serialized as camelCase?

sisou committed:
Format JSON for logs

[skip ci]

sisou committed:
Add timestamp to AppliedBlock logs

[skip ci]

nibhar commented on primitives/account/src/accounts.rs:
```diff
@@ -172,43 +191,48 @@ impl Accounts {
+        )?);
 
-        Ok(())
+        //TODO build batchinfo with results of previous reverts
```
Open TODO

nibhar commented on primitives/account/src/accounts.rs:
```diff
@@ -485,4 +541,35 @@ impl Accounts {
+            }
+        }
+
+        //assert tx_logs_recipients and tx_log_sender are over (?)
```
Good idea!

nibhar commented on primitives/account/src/staking_contract/staker.rs:
```diff
@@ -505,6 +534,17 @@ impl StakingContract {
             Some(x) => x,
         };
 
+        // If it was a no-op, we end right here.
```
Is there a reason this moved beneath the staker exist check?

nibhar commented on primitives/account/src/staking_contract/validator.rs:
```diff
@@ -236,12 +245,7 @@ impl StakingContract {
-            return Ok(());
-        }
-
+    ) -> Result<Vec<Log>, AccountError> {
```
I think this change to the return type could be applied to `revert_create_validator` as well. Makes for a more consistent interface and cleans up `primitives/account/src/staking_contract/traits.rs` a bit.

nibhar commented on primitives/account/src/staking_contract/validator.rs:
```diff
@@ -176,21 +177,24 @@ impl StakingContract {
         new_reward_address: Option<Address>,
         new_signal_data: Option<Option<Blake2bHash>>,
-    ) -> Result<UpdateValidatorReceipt, AccountError> {
+    ) -> Result<OperationInfo<UpdateValidatorReceipt>, AccountError> {
```
I think the return type for non revert operations can also be applied to `create_validator`. Same Reasoning as for the revert case.

nibhar commented on primitives/account/src/staking_contract/validator.rs:
```diff
@@ -252,6 +256,16 @@ impl StakingContract {
                 }
             };
+        // If it was a no-op, we end right here.
+        if receipt.no_op {
```
Again I wonder why this was moved beneath the validator exists check.

nibhar commented on primitives/account/src/staking_contract/staker.rs:
```diff
@@ -203,7 +215,7 @@ impl StakingContract {
         staker_address: &Address,
         value: Coin,
-    ) -> Result<(), AccountError> {
+    ) -> Result<Vec<Log>, AccountError> {
```
I think it would be more consistent to return 
```suggestion
    ) -> Result<AccountInfo, AccountError> {
```
as almost all other functions in this file do with the respective `*Info` type

nibhar commented on primitives/account/src/staking_contract/staker.rs:
```diff
@@ -35,7 +36,7 @@ impl StakingContract {
         value: Coin,
         delegation: Option<Address>,
-    ) -> Result<(), AccountError> {
+    ) -> Result<Vec<Log>, AccountError> {
```
I think this could also return 
```suggestion
    ) -> Result<AccountInfo, AccountError> {
```
Same reasoning as for `stake`.

sisou committed:
Fix data deserialization for create_basic_transaction_with_data

nibhar commented on primitives/account/src/staking_contract/traits.rs:
```diff
@@ -337,36 +342,50 @@ impl AccountTransactionInteraction for StakingContract {
         };
-
-        Ok(receipt)
+        acc_info.logs.insert(
```
Does the order of logs matter here or can this also be
```suggestion
        acc_info.logs.push(
```
?

If the order does matter, add a comment explaining why.

nibhar commented on primitives/account/src/staking_contract/traits.rs:
```diff
@@ -660,6 +714,15 @@ impl AccountInherentInteraction for StakingContract {
                             .current_lost_rewards
                             .remove(slot.slot as usize);
                     }
+                    logs.insert(
```
Does the order of logs matter here or can this also be
```suggestion
                    logs.push(
```
?

If the order does matter, add a comment explaining why.

sisou committed:
Unwrap blocks and validators in their RPC subscription streams

Co-Authored-By: Inês Cruz <ii.cruz@campus.fct.unl.pt>

nibhar commented on primitives/account/src/staking_contract/receipts.rs:
```diff
@@ -8,6 +8,9 @@ use nimiq_keys::{Address, PublicKey as SchnorrPublicKey};
 /// those inherents/transactions.
 
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct DefaultReceipt;
```
What is the purpose of this, is this used anywhere?

ii-cruz committed:
rpc client has one single function to subscribe to logs

jsdanielh reviewed (commented):
@ii-cruz Looks like there are several clippy warnings and an error. Could you address them?
You can check them from the last run of your code [here](https://github.com/nimiq/core-rs-albatross/pull/906/checks?check_run_id=7116450257).

jsdanielh commented on rpc-server/src/dispatchers/blockchain.rs:
```diff
@@ -529,4 +588,127 @@ impl BlockchainInterface for BlockchainDispatcher {
+
+    /// Subscribes to pre epoch validators events.
+    #[stream]
+    async fn election_validator_subscribe(
```
Maybe `validator_election_subscribe`?

jsdanielh commented on rpc-client/src/main.rs:
```diff
@@ -194,22 +205,50 @@ impl Command {
+            Command::FollowValidator { address } => {
+                let mut stream = client
+                    .blockchain
+                    .election_validator_subscribe(address)
```
Maybe `validator_election_subscribe`?

jsdanielh commented on rpc-interface/src/blockchain.rs:
```diff
@@ -97,10 +99,36 @@ pub trait BlockchainInterface {
 
     #[stream]
-    async fn head_subscribe(&mut self) -> Result<BoxStream<'static, Blake2bHash>, Self::Error>;
+    async fn election_validator_subscribe(
```
Maybe `validator_election_subscribe`?

jsdanielh commented on primitives/account/src/logs.rs:
```diff
@@ -0,0 +1,364 @@
+use crate::Receipt;
```
Mmm, it looks strange to me that we have this in `primitives/account` since this is not strictly related to accounts. Maybe consider moving it to one directory up (to `primitives/src`)?

ii-cruz replied:
I think it makes sense to place it in the same place as receipts, they are reasonably similar. Another argument for letting it stay here is to that the wrapper structs that are within logs are return types of many of the account methods.

jsdanielh commented on primitives/account/src/logs.rs:
```diff
@@ -0,0 +1,364 @@
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[cfg_attr(feature = "serde-derive", derive(serde::Serialize, serde::Deserialize))]
+// Renaming affects only the struct names and thus their tag, the "type" field.
+#[serde(rename_all = "kebab-case", tag = "type")]
```
Shouldn't this be `camelCase`? Maybe we can change it here and avoid adding it for all of the members?

sisou replied:
The name of the structs (which are renamed through this) are set as the 'type' property on the resulting objects. The string value should be kebab-cased IMO, this is on purpose.

This way it looks like this:
```
    type: 'applied-block',
```

And not like this:
```
    type: 'appliedBlock',
````

sisou replied:
Additionally, this rule only applies to direct members of the enum, not _their_ members in turn. That's why the rename rule needs to be added to all members individually anyway.

sisou replied:
If you find a way to apply it to all children fields as well, please tell!

jsdanielh replied:
Ok, yeah I see why it needs to be in the enum variants. But still it doesn't look consistent to me that this enum's tag would follow kebab case when everything else does camel case.

jsdanielh commented on primitives/account/src/logs.rs:
```diff
@@ -0,0 +1,364 @@
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[cfg_attr(feature = "serde-derive", derive(serde::Serialize, serde::Deserialize))]
+// Renaming affects only the struct names and thus their tag, the "type" field.
+#[serde(rename_all = "kebab-case", tag = "type")]
```
Shouldn't this be `camelCase`? Doesn't look consistent to me that we're renaming the members to use kebab-case?

jsdanielh commented on primitives/account/src/logs.rs:
```diff
@@ -0,0 +1,364 @@
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[cfg_attr(feature = "serde-derive", derive(serde::Serialize, serde::Deserialize))]
```
Aren't we missing `#[serde(rename_all = "camelCase")]`?

sisou replied:
It's unnecessary since I chose to rename the only field that would be affected with its own rename rule already.

jsdanielh replied:
I see your point but I'd still add it in case we add more members in the future and we forget to revise it

jsdanielh commented on primitives/account/src/logs.rs:
```diff
@@ -0,0 +1,364 @@
+
+// Batch Info is used as a return type of multiple transactions or batches applied to the blockchain.
+// It stores the transaction logs, inherent logs and receipts associated with the multiple transactions of the batch.
+#[derive(Default, Eq, PartialEq, Debug)]
```
Same here, aren't we missing `#[serde(rename_all = "camelCase")]`?

sisou replied:
AFAIK this struct is never sent as JSON.

jsdanielh commented on primitives/account/src/logs.rs:
```diff
@@ -0,0 +1,364 @@
+
+// Operation Info is used as a return type of some staker and validator related transactions.
+// It stores the transaction logs generated by the transaction and the receipts. The receipts can either be represented in bytes or in any receipt type defined.
+#[derive(Debug, PartialEq, Eq)]
```
Same here, aren't we missing `#[serde(rename_all = "camelCase")]`?

sisou replied:
No field would be affected by this rule, plus this struct is never sent as JSON AFAIK.

jsdanielh commented on primitives/account/src/logs.rs:
```diff
@@ -0,0 +1,364 @@
+}
+
+// Operation Info is used as a return type of some staker and validator related transactions.
+// It stores the transaction logs generated by the transaction and the receipts. The receipts can either be represented in bytes or in any receipt type defined.
```
Maybe consider splitting this line.

jsdanielh commented on primitives/account/src/logs.rs:
```diff
@@ -0,0 +1,364 @@
+}
+
+// Account Info is used as a return type of transactions to the blockchain and stores the logs and receipt associated with the transaction
+#[derive(Debug, PartialEq, Eq)]
```
Same here, aren't we missing `#[serde(rename_all = "camelCase")]`?

sisou replied:
No field would be affected by this rule, plus this struct is never sent as JSON AFAIK.

jsdanielh commented on primitives/account/src/logs.rs:
```diff
@@ -0,0 +1,364 @@
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[cfg_attr(feature = "serde-derive", derive(serde::Serialize, serde::Deserialize))]
+// Renaming affects only the struct names and thus their tag, the "type" field.
```
Mmm, it should affect all of the members according to the documentation: _Rename all the fields (if this is a struct) or variants (if this is an enum) according to the given case convention._

sisou replied:
It only renames the names of the variants, not their fields too.

jsdanielh replied:
Maybe it should be reworded as:
_Renaming affects only the enum variant names and its tag (the "type" field)_
?

jsdanielh commented on primitives/account/src/logs.rs:
```diff
@@ -0,0 +1,364 @@
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[cfg_attr(feature = "serde-derive", derive(serde::Serialize, serde::Deserialize))]
+// Renaming affects only the struct names and thus their tag, the "type" field.
```
Mmm, it should affect all of the members according to the documentation: _Rename all the fields (if this is a struct) or variants (if this is an enum) according to the given case convention._

sisou replied:
It does. It renames all _direct_ fields and variants.

jsdanielh replied:
Ok, I think I now understand the point of the comment, maybe it should be reworded as:
_Renaming affects only the enum variant names and its tag (the "type" field)_
?

nibhar commented on rpc-interface/src/types.rs:
```diff
@@ -692,6 +693,143 @@ impl Validator {
+}
+
+impl<T> BlockchainState<T> {
+    pub fn from_value(block_number: u32, block_hash: Blake2bHash, value: T) -> Self {
```
Small nitpick.
```suggestion
    pub fn new(block_number: u32, block_hash: Blake2bHash, value: T) -> Self {
```
But since this does not contain any logic and the fields are `pub`, you can also just use 

```rust
BlockchainState { block_number, block_hash, value }
```

nibhar reviewed (changes requested):
Looks good in general in my opinion. 

Clippy warnings and errors should for sure be fixed, and I would appreciate more documentation being added. 

Outside of that I commented mostly smaller nitpicks that I found reading this.

jsdanielh commented on rpc-interface/src/blockchain.rs:
```diff
@@ -97,10 +99,36 @@ pub trait BlockchainInterface {
+    ) -> Result<BoxStream<'static, BlockchainState<Validator>>, Self::Error>;
+
+    #[stream]
+    async fn logs_subscribe(&mut self) -> Result<BoxStream<'static, BlockLog>, Self::Error>;
```
```suggestion
    async fn logs_subscribe(&mut self) -> Result<BoxStream<'static, Log>, Self::Error>;
```
One more, shouldn't this be a `Log` instead of a `BlockLog`?

jsdanielh commented on rpc-interface/src/blockchain.rs:
```diff
@@ -97,10 +99,36 @@ pub trait BlockchainInterface {
+        &mut self,
+        addresses: Vec<Address>,
+        log_types: Vec<LogType>,
+    ) -> Result<BoxStream<'static, BlockLog>, Self::Error>;
```
Same here, shouldn't this be a `Log` instead of a `BlockLog`?

redmaner commented on rpc-interface/src/blockchain.rs:
```diff
@@ -97,10 +99,36 @@ pub trait BlockchainInterface {
+    ) -> Result<BlockchainState<Staker>, Self::Error>;
+
+    #[stream]
+    async fn head_subscribe(
```
From a RPC user perspective, I would like to propose some naming alternatives for the streaming methods:
* `head_subscribe` -> `subscribe_for_head_block`
* `head_hash_subscribe` -> `subscribe_for_head_block_hash`
* `election_validator_subscribe` -> `subscribe_for_validator_election_by_address` 
* `logs_subscribe` -> `subscribe_for_logs` or `subscribe_for_block_logs` or `subscribe_for_event_logs`
* `logs_by_type_and_addresses_subscribe` -> `subscribe_for_logs_by_addresses_and_types` or one of the other two variations mentioned above suffixed with `_by_addresses_and_types`

ii-cruz replied:
Thank you! 
The methods are renamed using those suggestions :)

redmaner commented:
First of all, great work @ii-cruz! Since I requested most of these features, I took the liberty of trying out your changes. I tested it quickly for now, and it looks great so far. I thinks this really adds some valuable ways to extract information that can be used by a lot of applications to great extends in the future. 

I added one comment to maybe improve the naming of the streaming methods. I prefer descriptive naming over short naming, and tried to propose a few alternatives that are more in line with the other methods. 

I agree with @nibhar on the documentation part. Especially for the filter method on logs it would be nice to know which types you could filter on for example. I think the logs methods are probably the most interesting to have documented, since the others are quite self explanatory.

ii-cruz committed:
Comments added. Merge tx_logs simplified. Renaming of some functions for clarity. No_op check on the reverts is back to being performed prior to any search for validator or staker. Standartization of return types for functions on staker and validator, all reverts return Vec<Log> and all commits return a OperationInfo.

ii-cruz committed:
Rename RPC server subscribe methods.
Added and corrected comments for clap.
Adjusted tx_logs merge function signature and comments for code clarity purposes.
Added clap casing for the LogType command.

ii-cruz committed:
Changing LogType to kebab-case

sisou reviewed (approved):
Just tested the latest state, both with `arpl` and with the built-in RPC client, and both work fine. The JSON sent is also correctly formatted, for the logs I tested. :+1:
LGTM :rocket:

nibhar reviewed (approved):
I'll create an issue for adding the documentation to our own rpc-client, but also to the server such that it is easily accessible. Also there is PR #923 related to this, for more consistency.
